gossip/src/duplicate_shred_handler.rs

```diff
@@ -144,18 +176,63 @@ impl DuplicateShredHandler {
             Some(SlotStatus::Frozen) => {
                 return false;
             }
-            Some(SlotStatus::UnfinishedProof(slot_map)) => {
-                if let Some(proof_chunkmap) = slot_map.peek(&data.from) {
-                    if proof_chunkmap.wallclock < data.wallclock {
-                        return false;
+            Some(SlotStatus::UnfinishedProof(slot_map)) => match slot_map.get(&data.from) {
+                None => {
+                    if slot_map.len() < MAX_PUBKEY_PER_SLOT {
+                        return true;
                     }
+                    return self.check_has_enough_stake_and_cleanup(&slot, &data.from);
                 }
-            }
+                Some(proof_chunkmap) => {
+                    return proof_chunkmap.wallclock >= data.wallclock;
+                }
+            },
             None => {}
         }
         true
     }
 
+    fn check_has_enough_stake_and_cleanup(&mut self, slot: &Slot, newkey: &Pubkey) -> bool {
+        match &mut self.cached_staked_nodes {
+            Some(cached_staked_nodes) => {
+                let newkey_stake = cached_staked_nodes.get(newkey).copied().unwrap_or_default();
+                // Ignore new pubkey without any stake.
+                if newkey_stake == 0u64 {
+                    return false;
+                }
+                let slot_map = self.chunk_map.get_mut(slot);
+                match slot_map {
+                    None => true,
+                    Some(SlotStatus::UnfinishedProof(chunk_map)) => {
+                        let mut heap = BinaryHeap::new();
+                        // Push new pubkey in first so it's compared with those in the table.
+                        heap.push(Reverse::<(u64, Pubkey)>((newkey_stake, *newkey)));
+                        for oldkey in chunk_map.keys() {
+                            let oldkey_stake =
+                                cached_staked_nodes.get(oldkey).copied().unwrap_or_default();
+                            heap.push(Reverse::<(u64, Pubkey)>((oldkey_stake, *oldkey)));
+                        }
+                        let mut newkey_popped = false;
+                        for _ in 0..(MAX_PUBKEY_PER_SLOT / 10) {
+                            if let Some(Reverse::<(u64, Pubkey)>((_, key))) = heap.pop() {
+                                if &key == newkey {
+                                    newkey_popped = true;
+                                } else {
+                                    chunk_map.remove(&key);
+                                }
+                            }
+                        }
+                        // If new key is popped off heap, then we should not insert this key because its
+                        // stake is too low, return false in this case.
+                        !newkey_popped
+                    }
+                    _ => false,
```

Comment:
Please avoid `_ => `.
Instead enumerate all possible cases.
If a new variant is added to `SlotStatus` enum then we get a compiler error to implement the right handling for the new variant here. However `_ =>` will silent that compiler error.

Reply:
done